cpp/src/arrow/python/extension_type.cc

```diff
@@ -71,6 +72,14 @@ PyObject* DeserializeExtInstance(PyObject* type_class,
 
 static const char* kExtensionName = "arrow.py_extension_type";
 
+std::string PyExtensionType::ToString() const {
+  std::stringstream ss;
+  OwnedRef instance(GetInstance());
+  ss << "extension<" << this->extension_name() << "<" << Py_TYPE(instance.obj())->tp_name
+     << ">>";
+  return ss.str();
+}
```

Comment:
This works fine, but there is one thing missing. Since `ToString` is called from arbitrary C++ code, the GIL needs to be taken. This is achieved by adding:
```c++
PyAcquireGIL lock;
```
at the beginning of this method definition.

Reply:
Amended commit to add `PyAcquireGIL`, thanks.